posthog/hogql/database/schema/channel_type.py

```diff
@@ -38,29 +56,130 @@ def create_initial_domain_type(name: str):
+                op=ast.CompareOperationOp.Eq,
+                left=expr,
+                right=ast.Constant(value=None),
+            )
+        ]
+        return ast.Or(exprs=exprs)
+    elif operator == CustomChannelOperator.ICONTAINS:
+        return ast.CompareOperation(
+            op=ast.CompareOperationOp.ILike,
+            left=expr,
+            right=ast.Constant(value=f"%{value}%"),
+        )
+    elif operator == CustomChannelOperator.NOT_ICONTAINS:
+        return ast.CompareOperation(
+            op=ast.CompareOperationOp.NotILike,
+            left=expr,
+            right=ast.Constant(value=f"%{value}%"),
+        )
+    elif operator == CustomChannelOperator.REGEX:
+        return ast.Call(
+            name="ifNull",
+            args=[
+                ast.Call(name="match", args=[ast.Call(name="toString", args=[expr]), ast.Constant(value=value)]),
+                ast.Constant(value=0),
+            ],
+        )
+    elif operator == CustomChannelOperator.NOT_REGEX:
+        return ast.Call(
+            name="ifNull",
+            args=[
+                ast.Call(
+                    name="not",
+                    args=[
+                        ast.Call(name="match", args=[ast.Call(name="toString", args=[expr]), ast.Constant(value=value)])
+                    ],
+                ),
+                ast.Constant(value=1),
+            ],
+        )
+
+    else:
+        raise NotImplementedError(f"PropertyOperator {operator} not implemented")
+
+
+def custom_rule_to_expr(custom_rule: CustomChannelRule, source_exprs: ChannelTypeExprs) -> ast.Expr:
+    conditions: list[Union[ast.Expr | ast.Call]] = []
+    for condition in custom_rule.conditions:
+        if condition.key == CustomChannelField.UTM_SOURCE:
+            expr = source_exprs.source
+        elif condition.key == CustomChannelField.UTM_MEDIUM:
+            expr = source_exprs.medium
+        elif condition.key == CustomChannelField.UTM_CAMPAIGN:
+            expr = source_exprs.campaign
+        elif condition.key == CustomChannelField.REFERRING_DOMAIN:
+            expr = source_exprs.referring_domain
+        else:
+            raise NotImplementedError(f"Property {condition.key} not implemented")
+        value = condition.value
+        if isinstance(value, list):
+            if len(value) == 0:
+                continue
+            elif len(value) == 1:
+                conditions.append(custom_condition_to_expr(expr, value[0], condition.op))
+            else:
+                conditions.append(
+                    ast.Call(name="or", args=[custom_condition_to_expr(expr, val, condition.op) for val in value])
+                )
+        else:
+            conditions.append(custom_condition_to_expr(expr, value, condition.op))
+    if len(conditions) == 0:
+        return ast.Constant(value=True)
+    elif len(conditions) == 1:
+        return conditions[0]
+    else:
+        return ast.Call(name=custom_rule.combiner.lower(), args=conditions)
```

Comment:
this is a really cool feature because it's super close to adding the specification pattern to our filters for arbitrary custom things not just channel types

that'd be scope creep here but so i'd be tempted to call this "specification" not "combiner" but that's not-blocking now